sockshop/catalogue/service.go

```diff
@@ -31,95 +32,157 @@ type Sock struct {
 // ErrNotFound is returned when there is no sock for a given ID.
 var ErrNotFound = errors.New("not found")
 
+// ErrDBConnection is returned when connection with the database fails.
+var ErrDBConnection = errors.New("database connection error")
+
+var baseQuery = "SELECT Sock.SockID AS id, Sock.name, Sock.description, Sock.price, Sock.count, Sock.ImageUrl_1, Sock.ImageUrl_2, GROUP_CONCAT(Tag.name) AS TagName FROM Sock JOIN SockTag ON Sock.SockID=SockTag.SockID JOIN Tag ON SockTag.TagID=Tag.TagID"
+
 // NewFixedService returns a simple implementation of the Service interface,
 // fixed over a predefined set of socks and tags. In a real service you'd
 // probably construct this with a database handle to your socks DB, etc.
-func NewFixedService(socks []Sock, tags []string) Service {
+func NewFixedService(db *sql.DB) Service {
 	return &fixedService{
-		socks: socks,
-		tags:  tags,
+		db: db,
 	}
 }
 
 type fixedService struct {
-	socks []Sock
-	tags  []string
+	db *sql.DB
 }
 
-func (s *fixedService) List(tags []string, order string, pageNum, pageSize int) []Sock {
+func (s *fixedService) List(tags []string, order string, pageNum, pageSize int) ([]Sock, error) {
 	var socks []Sock
-	{
-		socks = s.socks
-		socks = filter(socks, tags)
-		socks = sortBy(socks, order)
-		socks = cut(socks, pageNum, pageSize)
+	query := baseQuery
+
+	var args []interface{}
+
+	for i, t := range tags {
+		if i == 0 {
+			query += " WHERE Tag.name=?"
+			args = append(args, t)
+		} else {
+			query += " OR Tag.name=?"
+			args = append(args, t)
+		}
 	}
-	return socks
-}
 
-func (s *fixedService) Count(tags []string) int {
-	var socks []Sock
-	{
-		socks = s.socks
-		socks = filter(socks, tags)
+	query += " GROUP BY id"
+
+	if order != "" {
+		query += " ORDER BY ?"
+		args = append(args, order)
+	}
+
+	query += ";"
+	sel, err := s.db.Prepare(query)
+
+	if err != nil {
+		return []Sock{}, ErrDBConnection
 	}
-	return len(socks)
+	defer sel.Close()
+
+	rows, err := sel.Query(args...)
+	if err != nil {
+		return []Sock{}, ErrDBConnection
```

Comment:
Definitely useful, I'm wondering about potential security risk. Since this service was envisioned as being externally exposed returning the actual database error is potentially dangerous... I mean what would happen if hackers gained access to our vast quantities of sock data?!!?!